src/main/java/io/vertx/core/http/impl/WebSocketImplBase.java

```diff
@@ -107,29 +118,68 @@ public SocketAddress remoteAddress() {
 
   protected void writeMessageInternal(Buffer data) {
     checkClosed();
-    writePartialMessage(data, 0);
+    writeBinaryMessageChunks(data);
+  }
+
+  protected void writeTextMessageInternal(String text) {
+    checkClosed();
+    writeTextMessageChunks(text);
+  }
+
+  protected void writeBinaryMessageChunks(Buffer data) {
+    List<Buffer> chunks = chunkMessage(data);
+    writeChunksAsFrames(chunks, FrameType.BINARY);
+  }
+
+  protected void writeTextMessageChunks(String stringData) {
+    Buffer data = Buffer.buffer(stringData);
+    List<Buffer> chunks = chunkMessage(data);
+    writeChunksAsFrames(chunks, FrameType.TEXT);
   }
 
-  protected void writePartialMessage(Buffer data, int offset) {
-    int end = offset + maxWebSocketFrameSize;
-    boolean isFinal;
-    if (end >= data.length()) {
-      end  = data.length();
-      isFinal = true;
-    } else {
-      isFinal = false;
+  /**
+   * Splits the provided buffer into chunks of byte lengths not exceeding the maximum web socket frame size.
+   *
+   * @param buffer the buffer to be split. Is assumed to not be null
+   * @return list of chunks
+   */
+  private List<Buffer> chunkMessage(Buffer buffer) {
```

Comment:
Ok.  I thought this was a good chance to get rid of the recursive code, but not creating the list also makes sense.  I updated the pr.